engine/runtime-benchmarks/src/main/java/org/enso/interpreter/bench/benchmarks/semantic/MultiValueBenchmarks.java

```diff
@@ -0,0 +1,155 @@
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.Warmup;
+import org.openjdk.jmh.infra.BenchmarkParams;
+import org.openjdk.jmh.infra.Blackhole;
+
+/**
+ * These benchmarks compare performance of {@link EnsoMultiValue}. They create a vector in a certain
+ * configuration representing numbers and then they perform {@code sum} operation on it.
+ */
+@BenchmarkMode(Mode.AverageTime)
+@Fork(1)
+@Warmup(iterations = 3, time = 5, timeUnit = TimeUnit.SECONDS)
+@Measurement(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
+@OutputTimeUnit(TimeUnit.MILLISECONDS)
+@State(Scope.Benchmark)
+public class MultiValueBenchmarks {
+  private Value arrayOfNumbers;
+  private Value sum;
+  private Value self;
+  private final long length = 100000;
+
+  @Setup
+  public void initializeBenchmark(BenchmarkParams params) throws Exception {
+    var ctx = Utils.createDefaultContextBuilder().build();
+    var code =
+        """
+        from Standard.Base import Vector, Float, Number, Integer
+
+        type Complex
+            private Number re:Float im:Float
+
+        Complex.from (that:Number) = Complex.Number that 0
+
+        sum arr =
+            go acc i = if i >= arr.length then acc else
+                v = arr.at i
+                sum = acc + v
+                @Tail_Call go sum i+1
+            go 0 0
+
+        sum_re arr =
+            go acc i = if i >= arr.length then acc else
+                v = arr.at i . re
+                sum = acc + v
+                @Tail_Call go sum i+1
+            go 0 0
+
+        make_vector type n =
+            Vector.new n i->
+                r = 3 + 5*i
+                case type of
+                    0 -> r:Complex
+                    1 -> r:Integer
+                    2 -> r:Float
+                    3 ->
+                        c = r:Complex&Float
+                        c:Float
+                    4 ->
+                        c = r:Float&Complex
+                        c:Float
+                    5 -> r:Complex&Float
+                    6 -> r:Float&Complex
+        """;
+    var benchmarkName = SrcUtil.findName(params);
+    var src = SrcUtil.source(benchmarkName, code);
+    int type = Integer.parseInt(benchmarkName.substring(benchmarkName.length() - 1));
+
+    var module = ctx.eval(src);
+    this.self = module.invokeMember(MethodNames.Module.GET_ASSOCIATED_TYPE);
+    var makeVector = module.invokeMember("get_method", self, "make_vector");
+    this.arrayOfNumbers = makeVector.execute(self, type, length);
+    this.sum =
+        module.invokeMember(MethodNames.Module.EVAL_EXPRESSION, type == 0 ? "sum_re" : "sum");
+  }
+
+  /**
+   * The <b>base benchmark</b> for this suite. Measures how much it takes to access an Atom in a
+   * Vector, read {@code re:Float} field out of it and sum all of them together.
+   */
+  @Benchmark
```

Comment:
This is now the **base benchmark**. The results after are:
```
# the base one
MultiValueBenchmarks.sumOverComplexBaseBenchmark0         avgt    5  0.139 ± 0.013  ms/op
# these two are supposed to be faster
MultiValueBenchmarks.sumOverInteger1                      avgt    5  0.065 ± 0.003  ms/op
[info] MultiValueBenchmarks.sumOverFloat2                        avgt    5  0.073 ± 0.002  ms/op

# these should catch up with sumOverComplexBaseBenchmark0 one day
[info] MultiValueBenchmarks.sumOverComplexAndFloat5              avgt    5  8.580 ± 0.326  ms/op
[info] MultiValueBenchmarks.sumOverComplexFloatRecastedToFloat3  avgt    5  9.118 ± 0.483  ms/op
[info] MultiValueBenchmarks.sumOverFloatAndComplex6              avgt    5  8.110 ± 0.160  ms/op
[info] MultiValueBenchmarks.sumOverFloatComplexRecastedToFloat4  avgt    5  9.393 ± 0.648  ms/op
```
still 60 times slower than it should be.

Reply:
The current problem is in `reorderOnly` branch:

![reorderOnly](https://github.com/user-attachments/assets/d34cd4c6-b53a-4f01-baa2-2d07b96c13dc)

It seems to always allocate new arrays. Rather it should treat `EnsoMultiType` as compilation constants have everything ready from previous run.